### PR TITLE
Fix/2772

### PR DESCRIPTION
**Pull request by jcnelson**
This alters the mempool mining heuristic so as to consider high-fee transactions before low-fee transactions, and alters the mining behavior to try and mine a "small" block as soon as possible after discovering a new burnchain block or Stacks block, and then later on, try and mine a "big" block by RBF'ing the small block's commit later.  It achieves this by making both mempool iteration and block-building *interruptable*, so that once a soft deadline passes, the node will attempt to mine the transactions it had considered so far in the allotted time.

This is a draft for now, since it's undergoing live testing.

### CODE REVIEW

File: src/core/mempool.rs
Comment (gregorycoppola) [inline src/core/mempool.rs:500]: It seems to me like what you are trying to do here is:
1) get the origin addresses that are trying to spend the most
2) for those addresses, find the lowest nonces.

There is something to be said for the "simpler" strategy that you just sort by transaction fee.

For exmaple, to "cancel" a transaction in ETH, you submit a "0 transfer" transaction, with the same nonce, and a higher transaction fee than the transaction you want to cancel.

https://info.etherscan.com/how-to-cancel-ethereum-pending-transactions/

Maybe you can 1) sort by transaction fee, and then, to break ties 2) sort by nonce. This would support chaining.

But, if there is no tie on transaction fee, then don't involve the nonces, to allow transaction cancelation.

Users should have the awareness on their side to not submit the chained transaction (nonce = n + k) with a higher fee than the base transaction that is still to process (nonce = n).

**Review by gregorycoppola — COMMENTED** [review]
Hey Jude,

I just had a question about the sorting criterion...

File: src/core/mempool.rs
Comment (jcnelson) [reply src/core/mempool.rs:500]: None of the Hiro tooling currently supports RBF (ethereum's "cancel" operation), so if the logic only ordered by highest fee, your transactions could get stuck indefinitely under mempool congestion unless you're willing to manually change the fee of your transaction and re-sign and re-submit it (something only developers can be expected to do, and even they wouldn't be happy about it).  In this query, because the cumulative transaction fee for an origin is what gets used to prioritize transactions, you can use the current tooling to un-stuck your transactions simply by sending more of them.

**Comment by jcnelson** [comment]
Tests are currently failing due to an unrelated build error, seemingly brought by the `proc-macro-hack` package:

```
#10 9.489   = note: /usr/bin/ld: /usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libprofiler_builtins-3db8bb50be1aec25.rlib(InstrProfilingPlatformLinux.o): relocation R_X86_64_PC32 against undefined hidden symbol `__ehdr_start' can not be used when making a shared object
#10 9.489           /usr/bin/ld: final link failed: nonrepresentable section on output
#10 9.489           collect2: error: ld returned 1 exit status
#10 9.489           
#10 9.490 
#10 9.494 error: could not compile `proc-macro-hack` due to previous error
```

File: testnet/stacks-node/src/tests/neon_integrations.rs
Comment (kantai) [inline testnet/stacks-node/src/tests/neon_integrations.rs:1920]: Should these new tests be added to the integrations dockerfile?

File: testnet/stacks-node/src/config.rs
Comment (kantai) [inline testnet/stacks-node/src/config.rs:1115]: Can you document these new config options somewhere in this repo? Does the README.md make sense?

File: testnet/stacks-node/src/neon_node.rs
Comment (kantai) [inline testnet/stacks-node/src/neon_node.rs:185]: I think this makes more sense as a method of `BlockBuilderSettings`, e.g., `BlockBuilderSettings::from_config` or the like, and separated from the `neon_node` module.

File: testnet/stacks-node/src/neon_node.rs
Comment (kantai) [inline testnet/stacks-node/src/neon_node.rs:1364]: Does this change remove `sleep_before_tenure` entirely as a config option?

File: src/core/mempool.rs
Comment (kantai) [inline src/core/mempool.rs:475]: Should be a `debug`?

File: src/core/mempool.rs
Comment (kantai) [inline src/core/mempool.rs:759]: Should be a `debug`?

File: src/core/mempool.rs
Comment (kantai) [inline src/core/mempool.rs:778]: Should be a `debug`?

File: src/core/mempool.rs
Comment (kantai) [reply src/core/mempool.rs:500]: I think that using the raw fee of transactions, rather than the cumulative fee of chained transactions would be a better approach here. With the way it is currently written, this mining logic will _strongly_ favor the chattiest transaction submitters, such that if blocks are all full, you'd expect that any origin that only submits one transaction at a time will almost always get crowded out. I know that most tooling doesn't support RBF yet, but that is an easier change for wallets, libraries, etc. to roll out (and a feature that they'd want to support anyways), rather than miner block assembly, so I'd definitely advocate for using high fees.

File: testnet/stacks-node/src/neon_node.rs
Comment (jcnelson) [reply testnet/stacks-node/src/neon_node.rs:1364]: Yup

**Comment by kantai** [comment]
Can you add a `CHANGELOG.md` entry for this PR?

File: testnet/stacks-node/src/neon_node.rs
Comment (jcnelson) [reply testnet/stacks-node/src/neon_node.rs:185]: You can't `impl` a method for a struct in a different crate.

File: testnet/stacks-node/src/config.rs
Comment (jcnelson) [reply testnet/stacks-node/src/config.rs:1115]: Is there any mining documentation in the repo anywhere I can add to?

File: src/core/mempool.rs
Comment (jcnelson) [reply src/core/mempool.rs]: Done

File: src/core/mempool.rs
Comment (jcnelson) [reply src/core/mempool.rs]: Done

File: src/core/mempool.rs
Comment (jcnelson) [reply src/core/mempool.rs]: Done

File: testnet/stacks-node/src/neon_node.rs
Comment (jcnelson) [reply testnet/stacks-node/src/neon_node.rs:1364]: Removed

File: testnet/stacks-node/src/config.rs
Comment (jcnelson) [reply testnet/stacks-node/src/config.rs]: Added a section to the README

File: testnet/stacks-node/src/tests/neon_integrations.rs
Comment (jcnelson) [reply testnet/stacks-node/src/tests/neon_integrations.rs:1920]: Done

File: src/core/mempool.rs
Comment (jcnelson) [reply src/core/mempool.rs:496]: Yup; changed to prioritizing by just tx fee

**Comment by jcnelson** [comment]
> Can you add a CHANGELOG.md entry for this PR?

Added

File: testnet/stacks-node/src/neon_node.rs
Comment (kantai) [reply testnet/stacks-node/src/neon_node.rs]: Right -- I think this code shouldn't be in `neon_node` module (or stacks-node crate) -- it should live with the BlockBuilderSettings. But I guess that'd introduce a circular dependency on `Config`, so maybe just move this to `config.rs` and make it a member of `Config` like `config.make_block_builder_settings(attempt)`.

File: CHANGELOG.md
Comment (kantai) [inline CHANGELOG.md:8]: Change the subheader to either `## [Not Yet Released]` or `## [2.0.11.3.0]`

File: src/core/mempool.rs
Comment (kantai) [reply src/core/mempool.rs:496]: Can you change the iteration to no longer consider all of an origin's transaction as a block? This new ordering still would end up favoring chatty submitters (if I have a single high fee transaction, then all of my transactions would be considered for inclusion before considering anyone else's transactions).

Comment (jcnelson) [reply src/core/mempool.rs:496]: The downside to this is that it will lead to starvation if there are few distinct origin addresses (i.e. when the mempool isn't congested).  Maybe I can fix this too -- now that `iterate_candidates()` takes a `ClarityTx`, the miner ought to be able to do multiple mempool passes when filling a block, thereby mining multiple transactions from the same origin but only after all other origins have been considered.

Comment (kantai) [reply src/core/mempool.rs:496]: I think that's fine -- I'm more worried about the current behavior where a single origin address could monopolize the miner's attempts just by having one transaction be the highest fee tx.

Comment (jcnelson) [reply src/core/mempool.rs:496]: I made it so that all origin addresses that clear the minimum fee will be considered in a round-robin fashion, and that if time and space permits, the miner will continue to cycle through origin addresses until the block is filled.  I've added an integration test that verifies that multiple transactions from the same origin will be included in a block, but *only after* all origins have been considered first.  Sound good?

Comment (kantai) [reply src/core/mempool.rs:496]: Perfect!

**Review by kantai — APPROVED** [review]
This LGTM once the tests pass!

File: testnet/stacks-node/src/config.rs
Comment (gregorycoppola) [inline testnet/stacks-node/src/config.rs:1106]: The main part of this PR changes the ranking algorithm for the transactions.

Do we know that if you change that algorithm, you still need all these other changes?

**Review by gregorycoppola — APPROVED** [review]
LGTM because whatever you want to submit is fine with me.

However, I did have a question I am curious about. This PR does two things:
1) change the ranking algorithm for picking transactions
2) add settings that miners can configure

Do you know the relative contributions of the two? Would it be enough to just do (1)?

**Review by SyAsteria — APPROVED** [review]
LGTM. This should help clear things up in the mempool rn ⚡️

File: testnet/stacks-node/src/neon_node.rs
Comment (jcnelson) [reply testnet/stacks-node/src/neon_node.rs:1006]: Because I want the mock miner to try and re-try building better and better blocks, such as when new microblocks come in.

File: src/core/mempool.rs
Comment (jcnelson) [reply src/core/mempool.rs:145]: I think it's obvious from the codebase that it's microSTX.  Without exception, all of the variables we use to represent a quantity of STX are in units of microSTX.

File: testnet/stacks-node/src/neon_node.rs
Comment (jcnelson) [reply testnet/stacks-node/src/neon_node.rs:1629]: These two checks are sufficient.  If there is only one transaction in this block, and there has only been one block mined, then it means that the miner only ran tenure once and it resulted in an empty block (i.e. because it didn't have time to actually consider any transactions).

File: src/core/mempool.rs
Comment (jcnelson) [reply src/core/mempool.rs:550]: Good catch! No, we don't want that.

File: src/core/mempool.rs
Comment (jcnelson) [reply src/core/mempool.rs:546]: This method only does one pass through the origins. The callers in `miner.rs` repeatedly call `iterate_candidates()` with the same `epoch_tx`.  This way, if origin _O_ gets considered more than once, its expected nonces _n_ and _n+1_ will automatically be deduced by way of the inner `get_account()` call.

File: src/core/mempool.rs
Comment (jcnelson) [reply src/core/mempool.rs:1312]: Yes.  Eventually, we'll have a separate, _global_ minimum relay fee that will determine whether or not the transaction will be admitted to the mempool.  But it's important that this value is the same for all nodes; otherwise, we run the risk of transactions not getting well-propagated due to next-hop nodes having higher min-relay fees than the miners.

File: src/core/mempool.rs
Comment (jcnelson) [reply src/core/mempool.rs:437]: Fixed

File: src/core/mempool.rs
Comment (jcnelson) [reply src/core/mempool.rs]: Fixed

File: src/core/mempool.rs
Comment (jcnelson) [reply src/core/mempool.rs]: Fixed

File: src/core/mempool.rs
Comment (gregorycoppola) [reply src/core/mempool.rs:145]: I think it makes sense to use context to save on comments, but I would agree with @pavitthrap in the sense that I didn't realize this was assumed until this discussion.

**Comment by jcnelson** [comment]
The atlas integration test repeatedly passes locally for me.  Can I just merge this without that specific test passing?

**Comment by kantai** [comment]
> The atlas integration test repeatedly passes locally for me. Can I just merge this without that specific test passing?

The atlas integration test does appear to just be flaky in GH actions -- it's passed at different times for this PR. The sampled_genesis tests, however, don't appear to just be flaky-- they have not all passed yet for this PR.

File: src/core/mempool.rs
Comment (pavitthrap) [reply src/core/mempool.rs:543]: Looking at the code, I see that iterate_candidates is called by `build_anchored_block` and `mine_next_microblock`. 

In `build_anchored_block`, I see that `epoch_tx` is initialized, `iterate_candidates` is called once at the top level of the function, and afterwards, `epoch_finish` is called. I'm confused how `iterate_candidates` is being called repeatedly?

**Pull request merged**
